source/slang/slang-ir-metadata.cpp

```diff
@@ -0,0 +1,77 @@
+// slang-ir-metadata.cpp
+#include "slang-ir-metadata.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+// This file currently implements a pass that collects information about the shader parameters that
+// are referenced in the IR. It's named 'metadata' in order to support other potential code
+// analysis scenarios in the future.
+
+
+// Inserts a single resource binding (which takes `count` slots, where 0 means unbounded) into the list of resource ranges.
+static void _insertBinding(List<ShaderBindingRange>& ranges, LayoutResourceKind kind, UInt spaceIndex, UInt registerIndex, UInt count)
+{
+    // Construct a new range from the provided resource.
+    ShaderBindingRange newRange;
+    newRange.category = kind;
+    newRange.spaceIndex = spaceIndex;
+    newRange.registerIndex = registerIndex;
+    newRange.registerCount = count;
+
+    // See if the new range is adjacent to any of the existing ranges, merge with that.
+    for (auto& range : ranges)
+    {
+        if (range.adjacentTo(newRange))
+        {
+            range.mergeWith(newRange);
+            return;
+        }
+    }
```

Comment:
Does this logic handle the case where two previously non-adjacent ranges become adjacent as a result of the range between them being added? That is, if we add `[0,A)` then `[B,C)` and finally `[A, B)` (with `A < B < C`), will the result have the single range `[0, C)`? It doesn't seem like it from this merge logic.

Reply:
No, there will be two ranges. We can of course run a compaction pass on the ranges after all of them have been collected, but is that necessary?